DeviceCommunications/SerialComsStream.cs

```diff
@@ -0,0 +1,123 @@
+﻿using DeviceCommunications.Interfaces;
+using System;
+using System.IO.Ports;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace DeviceCommunications
+{
+    public class SerialComsStream : IDeviceComsStream
+    {
+        private SerialPort _SerialPort = new SerialPort();
+
+        private CancellationTokenSource _cts;
+
+        private object _RxDataLock = new object();
+        private bool _RxDataRead;
+
+
+        // Class functions
+        public SerialComsStream(string portName, int baudRate, Parity parity, int dataBits, StopBits stopBits)
+        {
+            _SerialPort = new SerialPort(portName, baudRate, parity, dataBits, stopBits);
+            _SerialPort.DtrEnable = true;
+            _SerialPort.RtsEnable = true;
+        }
+
+        private void MonitorForReceivedData(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                lock (_RxDataLock)
+                {
+                    if  (
+                            (false == _RxDataRead)
+                            && (0 != _SerialPort.BytesToRead)
+                        )
+                    {
+                        if (null != DataReceivedEvent)
+                        {
+                            DataReceivedEvent(this, EventArgs.Empty);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Interface functions
+
+        public event EventHandler DataReceivedEvent;
+
+        bool IDeviceComsStream.IsConnected
+        {
+            get
+            {
+                return _IsConnected;
+            }
+        }
+
+        private bool _IsConnected
+        {
+            get { return (bool)_SerialPort.IsOpen; }
+        }
+
+        bool IDeviceComsStream.Connect()
+        {
+            if (_IsConnected)
+            {
+                _SerialPort.Close();
+            }
+
+            _SerialPort.Open();
+
+            if (_IsConnected)
+            {
+                // now that we are connected, make sure we monitor for received data
+                _cts?.Cancel();
+                _cts = new CancellationTokenSource();
+                _RxDataRead = false;
+                Task.Run(() => MonitorForReceivedData(_cts.Token), _cts.Token);
+            }
```

Comment:
This can totally work, but I've found with C#'s async framework it can work out a bit better to make things 'async all the way up', which results in returning the task that's monitoring the serial port to the caller. It ends up being an awkwardly named function like 'ConnectAndMonitor' or something like that, but it makes error handling a lot easier and lets us use the Serial Stream's own async read methods.
(BTW - It might be better now, but I was warned by a technical mentor that System.IO.Ports.SerialPort is terrible in many ways by way of this article: https://www.sparxeng.com/blog/software/must-use-net-system-io-ports-serialport
which also has some good suggestions for workarounds)

Reply:
Using C#s async functionality seems good, this isn't my wheelhouse being a firmware but it sounds like a good idea, I've never had issues with System.IO.Ports.SerialPort but I've never done anything hardcore with it.

Reply:
I'll have a read of the article but it is a lot easier than using the serial port in a win32 app in C++.

Reply:
> Fair enough! I haven't either, but similarly wasn't doing anything too taxing